[MRG] Cemoody/bhtsne Barnes-Hut t-SNE

cemoody opened the pull request:
## Introduction

This PR presents the Barnes-Hut implementation of t-SNE. t-SNE is used to visualize high-dimensional data in a low dimensional space that attempts preserve the pairwise high-dimensional similarities in a low-dimensional embedding. The Barnes-Hut algorithm, which is used by astrophysicists to perform N-body simulations, allows the calculation of the t-SNE embedding in O(NlogN) time instead of O(N^2). 

This effectively allows us to learn embeddings of data sets with millions of elements instead of tens of thousands.
## Example

An IPython Notebook is available where we've put together simple install instructions & a demo of both methods:
http://nbviewer.ipython.org/urls/gist.githubusercontent.com/cemoody/01135ef2f26837548360/raw/76ce7f0bac916a516501ead719b513b22430cad0/Barnes-Hut%20t-SNE%20Demo.ipynb

![bhtsne](https://cloud.githubusercontent.com/assets/3419930/5566392/24d8b7dc-8edb-11e4-9e67-4f319fc6dc59.png)
## Performance

This compares the timings for the Barnes-Hut approximation against the current method (note the log y-axis).

![timing](https://cloud.githubusercontent.com/assets/3419930/5566592/9f8ba956-8ee2-11e4-88e9-2504485f06c2.png)

The following diagram shows the relative speedup between both versions:
![speedup](https://cloud.githubusercontent.com/assets/3419930/5566360/f4802f94-8ed9-11e4-891c-da999a0426e8.png)
## TODO
- [x] Create fit & transform methods to update using new data
- [x] Add test for transform called before fit
- [x] Add test for skip_num_points
- [x] Raise error if method='barnes_hut' and using sparse data
- [x] Raise error if method='barnes_hut' and embedding dim > 3 
- [x] Transform 64bit input floats to 32bit
- [x] Update the narrative documentation in doc/modules/manifold.rst
- [x] Add toy data to test perplexity
- [x] Fix failing tests
- [x] Expose tree consistency checks as unit tests
- [x] Fix memory leaks
- [x] Segfaults on data larger than 50k elements
- [x] PEP8 the code
- [x] Include usage documentation
- [x] Remove extra imports, print statements, pdb statements
- [x] Ensure python 3 works
- [x] Ensure output dimensionality works for 2D or 3D
- [x] Am I using memviews or returning full arrays appropriately?
- [x] Incorporate into SKLearn
- [x] Remove GIL as much as possible in Cython code
- [x] Add answer tests
- [x] Ensure old t-SNE tests works
- [x] Changed perplexity calculation to nearest neighbors
- [x] Changed positive gradient calc to nearest neighbors
## Learn
- You can read more about the technique in general here:

http://lvdmaaten.github.io/tsne/
- The Barnes-Hut approximation is here:

http://lvdmaaten.github.io/publications/papers/JMLR_2014.pdf

arjoly committed:
ENH add a benchmark on mnist

GaelVaroquaux commented:
Hey Christopher,

Thanks a lot for contributing this. I haven't had time to review it, but
it is exciting! I think that the Barnes-Hut t-SNE algorithm is very
important to us to scale up.

cemoody commented:
Thanks @GaelVaroquaux and @ogrisel for your comments! @nicktrav and I are slowly working through our TODO list and all of @ogrisel's wonderful sugestions (thank you!) -- we're hoping to be ready for a fuller review in the coming week or two.

AlexanderFabisch commented on sklearn/manifold/t_sne.py:
```diff
@@ -111,17 +152,135 @@ def _kl_divergence(params, P, alpha, n_samples, n_components):
+    Uses Barnes-Hut tree methods to calculate the gradient that
+    runs in O(NlogN) instead of O(N^2)
+
+    Parameters
```
the documentation of neighbors and verbose is missing here, too

AlexanderFabisch commented on sklearn/manifold/t_sne.py:
```diff
@@ -433,24 +639,51 @@ def _fit(self, X):
             else:
                 distances = pairwise_distances(X, metric=self.metric)
 
+        assert np.all(distances >= 0), "All distances should be positive"
```
you should probably raise a ValueError in this case, either the metric or the distances given as X are not correct

cemoody commented:
Hi @GaelVaroquaux, @ogrisel & @AlexanderFabisch thanks for the comments! 11 hours ago I thought this was going to take a week or two, but I believe we're now ready for review :) I'll try to respond to everyone's comments all at once.

- We've removed all OpenMP requirements -- the overhead of launching threads was bigger than the tradeoff, even for large datasets with 10k elements. Once we've figured out building scikit-learn with OpenMP we can revisit this question, but let's table the issue for the time being.
- We've modified the narrative docs to reflect the new performance and a extra parameters. The change is minimal since the algorithm performs more or less the same, just faster.
- Added toy data testing & comparing the perplexity of the 2 methods.

The rest of the changes we've immediately incorporated. Tests should now pass too, which I think that means this is ready for merge review, so I've changed the title to [MRG].

cc @nicktrav

ogrisel commented on sklearn/manifold/_debugvars.h:
```diff
@@ -0,0 +1,3 @@
+#define DEBUGFLAG 0
```
Is it really necessary to introduce a new file just for this? As `_debugvars.h` is only included in `_barnes_hut_tsne.pyx` I would rather put a an inline definition of that constant. AFAIK there is no Cython keyword to default a `const` variable but maybe a `cdef enum` would do. It needs to be bench to make sure that it does not cause a performance regression but I think it should be fine.

cemoody replied:
Yeah, including an extra file just for compile-time debug flags is gnarly. Didn't think about enums before now, but using an enum is a much more elegant solution, and performance-wise seems to make no (significant) impact. Thanks -- I've made this change.

AlexanderFabisch commented on doc/modules/manifold.rst:
```diff
@@ -513,12 +514,16 @@ divergence could increase during this phase. Usually it does not have to be
+but less accurate results. 
 
-Standard t-SNE that has been implemented here is usually much slower than
+The Barnes-Hut t-SNE that has been implemented here is usually much slower than
```
Is it really still slower than other manifold learning algorithms?

cemoody replied:
Yeah, at least for the toy datasets: the speed up is only ~2-3x on the 2000 digits examples, which still makes it the slowest. Perhaps it's competitive with the next-slowest, MDS?

ogrisel replied:
On my box, `TSNE` (with the default Barnes Hut approximation) is always slightly faster than our implementation of `MDS` in all the `examples/manifold` scripts. Still it's slower when compared to other manifold methods.

ogrisel replied:
I would replace "is usually much slower than" by "can be significantly slower than".

ogrisel commented on sklearn/manifold/t_sne.py:
```diff
@@ -48,15 +52,55 @@ def _joint_probabilities(distances, desired_perplexity, verbose):
     # the desired perplexity
     conditional_P = _utils._binary_search_perplexity(
-        distances, desired_perplexity, verbose)
+        distances.astype(np.float), desired_perplexity, verbose)
```
This will trigger a memory copy even if `distances` already has the right type. Please use `sklearn.utils.fixes.astype` and use it with `astype(distances, np.float, copy=False)`. Also I would rather use an explicit precision with either `np.float32` or `np.float64` or better a dtype that reflect the precision of the input data instead of an arbitrary yet platform dependent `np.float` dtype.

ogrisel replied:
Maybe `sklearn.utils.validation.as_float_array` might be what you want here actually.

AlexanderFabisch commented on sklearn/manifold/t_sne.py:
```diff
@@ -477,29 +758,54 @@ def _tsne(self, P, alpha, n_samples, random_state, X_embedded=None):
                                                    self.n_components)
         params = X_embedded.ravel()
 
+        kwargs = {}
```
I would write

```python
opt_args = {"n_iter": 50, "momentum": 0.5, "learning_rate": self.learning_rate, "verbose": self.verbose,
            "it": 0, "n_iter_check": 25, "kwargs": dict(skip_num_points=skip_num_points)}
```

ogrisel commented on sklearn/manifold/t_sne.py:
```diff
@@ -111,17 +160,147 @@ def _kl_divergence(params, P, alpha, n_samples, n_components):
+        gradient function
+
+    alpha : float
+        Degrees of freedom of the Student's-t distribution.
```
I think it should be named `degrees_of_freedom` to make the code more readable but as this is private API this is no big deal.

GaelVaroquaux replied:
> I think it should be named degrees_of_freedom to make the code more readable

+1

> but as this is private API this is no big deal.

Still, it would make the code more readable.

AlexanderFabisch commented on sklearn/manifold/t_sne.py:
```diff
@@ -358,6 +555,25 @@ class TSNE(BaseEstimator):
+    method : string (default: 'barnes_hut')
+        By default the gradient calculation algorithm uses Barnes-Hut
+        approximation running in O(NlogN) time. method='standard' will use
+        will run on the slower, but exact, algorithm in O(N^2) time. The
```
will use or will run?

AlexanderFabisch commented on sklearn/manifold/t_sne.py:
```diff
@@ -403,17 +625,42 @@ def __init__(self, n_components=2, perplexity=30.0,
+            raise ValueError("'angle' must be between 0.0 - 1.0")
+        if self.method == 'barnes_hut' and self.n_components > 3:
+            raise ValueError("method='barnes_hut' only available for "
+                             "n_components < 4.")
```
You should mention that in the docstring of the constructor.

AlexanderFabisch commented:
Could you add the performance comparison to `examples/manifold` and make sure that it does not take much longer than 10^1 seconds (`n_samples=1000-2000`)?

ogrisel commented on doc/modules/manifold.rst:
```diff
@@ -502,6 +502,7 @@ three parameters that control the optimization of t-SNE:
 * early exaggeration factor
 * learning rate
 * maximum number of iterations
+* angle
```
The paragraph above needs an update: "there are three" => "there are four".

AlexanderFabisch replied:
Actually, `angle` does not really influence the optimization. It controls the approximation of the gradient. I would remove it completely and explain it in another paragraph.

ogrisel replied:
As is an approximation of the true cost: while the default value should of 0.5 or lower should not impact the results much, setting it to a high value should impact the quality of the embedding.

ogrisel replied:
Maybe `optimization of t-SNE` should be rephrased `optimization of t-SNE and therefore possibly the quality of the resulting embedding`

cemoody replied:
Yep, agreed that high values of `angle` should negatively impact the quality of the embedding, and so is as relevant as the learning rate or number of iterations.

AlexanderFabisch commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -233,6 +278,87 @@ def test_pca_initialization_not_compatible_with_precomputed_kernel():
+                          [0]])
+    grad_output = np.array([[-2.37012478e-05, -6.29044398e-05],
+                            [2.37012478e-05, 6.29044398e-05]])
+    yield _run_answer_test, pos_input, pos_output, neighbors, grad_output
```
I think this is more difficult to understand than `_run_answer_test(pos_input, pos_output, neighbors, grad_output)`

ogrisel replied:
+1 for unfolding this test and getting rid of the yield expression.

AlexanderFabisch commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -263,3 +389,90 @@ def test_chebyshev_metric():
+
+
+def test_no_4D_on_barnes_hut():
+    """No sparse matrices allowed on Barnes-Hut"""
```
This description is not correct

ogrisel replied:
The docstring is does not reflect the test. As the test method name is explicit there is no need for a docstring anyway.

AlexanderFabisch commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,626 @@
+# cython: boundscheck=False
```
I have never seen that before. Does it do the same as `@cython.boundscheck(False)` for each method?

cemoody replied:
Yup! Much more concise than having a decorator on every single function.

AlexanderFabisch commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,626 @@
+# Author: Christopher Moody <chrisemoody@gmail.com>
+# Author: Nick Travers <nickt@squareup.com>
+import numpy as np
+from cython.parallel cimport prange, parallel
```
These are not required

AlexanderFabisch commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,626 @@
+from cython.parallel cimport prange, parallel
+from libc.stdio cimport printf
+cimport numpy as np
+cimport cython
```
This does not seem to be required, either?

ogrisel commented on doc/modules/manifold.rst:
```diff
@@ -513,12 +514,16 @@ divergence could increase during this phase. Usually it does not have to be
-is the number of output dimensions and :math:`N` is the number of samples.
+and the computation of the gradient is on :math:`O[d N log(N)]`, where :math:`d`
+is the number of output dimensions and :math:`N` is the number of samples. This
+improves on the standard t-SNE which is :math:`O[d N^2]`. 
```
This paragraph does not make it explicit that both methods are implemented in scikit-learn. I think it's important to state the pro and the cons of each method. In particular:

- BH is significantly more scalable In practice BH can be used to embed hundred of thousand of data points while the standard can barely handle thousands of samples,
- BH only works when the target dimensionality is 3 or less (typically 2 for a visualization with a matplotlib scatter plot),
- BH is an approximation of the standard method. The approximation is parameterized with the `angle` parameter, therefore the `angle` parameter is unused when `method="standard"`
- BH only works with dense input data. Sparse data matrices can only be embedded with the standard method.

For visualization purpose (which is the main use case of t-SNE), using the BH method is strongly recommended. The standard t-SNE method is useful for checking the theoretically properties of the embedding possibly in higher dimensional space but limit to small datasets due to computational constraints.

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,626 @@
+        if (pos[ax] == node.cur_pos[ax]) & (index != node.point_index):
+            offset[ax] = 1 - offset[ax]
+            if DEBUGFLAG:
+                printf("[t-SNE] flipped node.cur_pos %i [%f, %f] pos %i [%f, %f]\n", node.point_index, node.cur_pos[0], node.cur_pos[1], index, pos[0], pos[1])
```
style: please add a line break:

```python
                 printf("[t-SNE] flipped node.cur_pos %i [%f, %f] pos %i [%f, %f]\n",
                        node.point_index, node.cur_pos[0], node.cur_pos[1], index, pos[0], pos[1])
```

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,626 @@
+    for ax in range(dimension):
+        root.cum_com[ax] *= frac_seen
+        if (pos[ax] > (root.le[ax] + root.w[ax])):
+            printf("[t-SNE] Error: point (%1.9e) is above right edge of node (%1.9e)\n", pos[ax], root.le[ax] + root.w[ax])
```
same comment as above: try to respect the 80 columns limit when it's easy to do.

AlexanderFabisch replied:
pep8 gives me 167 warnings for this file, I think this could be reduced significantly. Although we can ignore something like `_barnes_hut_tsne.pyx:458:14: E225 missing whitespace around operator` these are important:

* `_barnes_hut_tsne.pyx:192:80: E501 line too long (82 > 79 characters)`
* `_barnes_hut_tsne.pyx:407:47: E231 missing whitespace after ','`
* `_barnes_hut_tsne.pyx:442:1: E303 too many blank lines (3)`
* `_barnes_hut_tsne.pyx:481:46: W291 trailing whitespace`
* ...

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -263,3 +389,90 @@ def test_chebyshev_metric():
+    X = random_state.randn(100, 2)
+    tsne = TSNE(n_components=2, perplexity=2, learning_rate=100.0,
+                random_state=0, method='barnes_hut')
+    m = ".*Cannot call `transform` unless `fit` has.*"
```
I think the `.*` are not needed.

ogrisel commented on sklearn/manifold/t_sne.py:
```diff
@@ -403,17 +625,42 @@ def __init__(self, n_components=2, perplexity=30.0,
+        if self.method == 'barnes_hut' and sp.issparse(X):
+            raise TypeError('A sparse matrix was passed, but dense '
+                            'data is required. Use X.toarray() to '
+                            'convert to a dense numpy array or '
```
"if the shape is small enough for it to fit in memory"

ogrisel replied:
Maybe it would be smarter to recommend to transform the sparse data into a 50 to 100 dimensional sparse using the `TruncatedSVD` transformer for instance.

AlexanderFabisch replied:
That makes sense. I think that should be documented in the docstring.

ogrisel replied:
Indeed, +1 for putting this tip both in the docstring and the exception message.

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -110,6 +116,39 @@ def test_binary_search():
+        P1top = P1.ravel()[idx][:topn]
+        P2top = P2k.ravel()[idx][:topn]
+        assert_array_almost_equal(P1top, P2top, decimal=2)
+
```
Maybe it would be interesting to add a test checks that:

`TSNE(method='standard')` yields an embedding with a perplexity or 1-NN query accuracy similar to one generated with `TSNE(method='barnes_hut', angle=0)`.

ogrisel commented:
If think the Barnes Hut t-SNE paper tsne paper mentions that the data can typically be preprocessed with a PCA with truncated to the top 50 components. I am wondering if this is equivalent to our `init="pca"` of if it's equivalent to:

```python
from sklearn.decomposition import TSNE, RandomizedPCA
from sklearn.pipeline import make_pipeline

pca_preprocessed_tsne = make_pipeline(RandomizedPCA(n_components=50), TSNE())
pca_preprocessed_tsne.fit_transform(data)
```

If it is the later I think we should either document this usage in the narrative documentation or better in an example.

ogrisel commented:
BTW @cemoody if you use sklearn as part of your work at squareup.com, it would be great to submit another PR to explain how you do so in https://github.com/scikit-learn/scikit-learn/blob/master/doc/testimonials/testimonials.rst to appear here: http://scikit-learn.org/stable/testimonials/testimonials.html

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -233,6 +278,87 @@ def test_pca_initialization_not_compatible_with_precomputed_kernel():
+        set of children.
+
+        These tests & answers have been checked against
+        the reference implementation by LvdM
```
Please follow the PEP 257  convention for docstring formatting.

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -263,3 +389,90 @@ def test_chebyshev_metric():
+        counts = np.zeros(3, dtype='int64')
+        _barnes_hut_tsne.check_quadtree(X, counts)
+        m = "Tree consistency failed: unexpected number of points at root node"
+        assert counts[0] == counts[1], m
```
Please use `assert_equal` to get informative failure message with the expected and observed count values.

AlexanderFabisch commented:
@ogrisel Transforming the original data with PCA is not equivalent to initializing the embedded space with the PCA transformed original data.

AlexanderFabisch commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,625 @@
+cimport cython
+
+# Implementation by Chris Moody & Nick Travers
+# original code by Laurens van der Maaten
```
Maybe we should make clear that this is your work and replace

```python
# original code by Laurens van der Maaten
# for reference implemenations and papers describing the technique:
# http://homepage.tudelft.nl/19j49/t-SNE.html
```
by
```python
# See http://homepage.tudelft.nl/19j49/t-SNE.html for reference
# implementations and papers describing the technique
```

ogrisel replied:
+1

AlexanderFabisch commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,625 @@
+# It allows us to write printf debugging lines
+# and remove them at compile time
+cdef enum:
+    DEBUGFLAG = 0
```
An alternative to this would be:

```python

    config.add_extension("_barnes_hut_tsne",
                         sources=["_barnes_hut_tsne.c"],
                         include_dirs=[numpy.get_include()],
                         libraries=libraries,
                         extra_compile_args=["-O3", "-DDEBUGFLAG=0"]) <---
```

ogrisel replied:
Good idea.

cemoody replied:
Does this actually work?

If I remove entirely:
```python
cdef enum:
    DEBUGFLAG = 0
```

And add in setup.py:

```python
...extra_compile_args=["-O3", "-DDEBUGFLAG=1"])...
```
Then Cython complains that DEBUGFLAG isn't a declared variable (`Accessing Python global or builtin not allowed without gil`) -- which of course it isn't. 

So then declaring the DEBUGFLAG but leaving it uninitialized (`def int DEBUGFLAG`) compiles, but doesn't seem to respect the compile args (`-DDEBUGFLAG=1`). Only if I initialize `cdef int DEBUGFLAG=1` does it work. 

Would love guidance here. I've left this using:
```python
cdef enum:
    DEBUGFLAG = 0
```
for the time being.

AlexanderFabisch replied:
Sorry, I did not try to generate the cython file. :)

AlexanderFabisch commented on sklearn/manifold/_utils.pyx:
```diff
@@ -109,3 +109,111 @@ cpdef np.ndarray[np.float_t, ndim=2] _binary_search_perplexity(
+
+
+@cython.boundscheck(False)
+cpdef np.ndarray[np.float_t, ndim=2] _binary_search_perplexity_nn(
```
There are only 6 lines that have been changed in comparison to `_binary_search_perplexity`. It should be possible to combine both functions.

ogrisel commented:
> @ogrisel Transforming the original data with PCA is not equivalent to initializing the embedded space with the PCA transformed original data. Transforming the input changes the objective function. The initialization does not change it.

Yes I agree. It's just that I don't remember which of the 2 options is explicitly recommended in the Barnes Hut t-SNE paper.

AlexanderFabisch commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,627 @@
+cimport numpy as np
+import numpy as np
+
+cdef extern from "math.h":
```
Is there a reason why you don't use
```python
from libc.math cimport sqrt
```
here?

amueller commented on sklearn/manifold/_utils.pyx:
```diff
@@ -2,33 +2,37 @@ from libc cimport math
+    """Binary search for sigmas of conditional Gaussians. 
+    
+    This approximation reduces the computational complexity from O(N^2) to O(uN). 
+    See the exact method '_binary_search_perplexity' for more details.
```
I don't understand this comment. Where is the exact method?

amueller commented on sklearn/manifold/t_sne.py:
```diff
@@ -374,10 +597,10 @@ class TSNE(BaseEstimator):
-           [ -714.79...,  3243.34...],
-           [  957.30..., -2505.78...],
-           [-1130.28...,  -974.78...])
+    array([[ 181.10080056,  -43.81465995],
```
doctest fails :-/

amueller commented on sklearn/manifold/t_sne.py:
```diff
@@ -521,5 +836,54 @@ def fit_transform(self, X):
+        The new gradient is calculated using contributions from
+        previously fit data, but only the new data is transformed.
+        This is not the equivalent of running fit_transform,
+        since calling fit(X) followed by transform(X) runs the
```
This does not make me super happy, but I'm not sure if there is an alternative :-/

amueller replied:
Oh, I see you raise a warning. That is fair enough then.

cemoody replied:
I'm not a huge fan of this behavior either, but agreed, there is no alternative. I think the best we can do is issue a warning?

amueller replied:
I agree.

amueller commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -99,17 +103,55 @@ def flat_function(_):
+    """Binary perplexity search approximation.
+
+    Should be approximately equal to the slow method when we use
+    all points as nighbors.
```
neighbors

amueller commented:
@ogrisel did you review the math?

arjoly committed:
ENH Improve script output display

cemoody commented:
The [CI tests (see line 1603)](https://travis-ci.org/scikit-learn/scikit-learn/jobs/46747120#L1603) are failing on a piece of code seemingly tangentially connected to TSNE and I cannot diagnose why:

`FAIL: sklearn.tests.test_common.test_all_estimators('TSNE', <class 'sklearn.manifold.t_sne.TSNE'>)`

because:

` '<mtrand.RandomState object at 0x15f8810> != None' = self._formatMessage('<mtrand.RandomState object at 0x15f8810> != None', '<mtrand.RandomState object at 0x15f8810> != None')`

I also can't recreate this on my machine with nosetests, so I'm guessing this is an automatically generated test that is now only failing on CI?

ogrisel commented:
You should be able to reproduce the failing test by running the full test suite (or at least `nosetests sklearn/tests/test_common.py`).

ogrisel commented on sklearn/manifold/t_sne.py:
```diff
@@ -402,19 +632,50 @@ def __init__(self, n_components=2, perplexity=30.0,
         self.init = init
         self.verbose = verbose
-        self.random_state = random_state
+        self.random_state = check_random_state(random_state)
```
This is the reason of the failure. This should be:

```python
self.random_state = random_state
```

as the `__init__` method should mutate the hyperparameters.

And in `fit` turn it into an RNG in a local variable:

```python
random_state = check_random_state(random_state)
```

ogrisel commented:
> @ogrisel did you review the math?

Not yet.

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -262,3 +422,126 @@ def test_chebyshev_metric():
+                                  perplexity, False)
+    kl, gradbh = _kl_divergence_bh(params, Pbh, neighbors_nn, degrees_of_freedom, n_samples,
+                      n_components, angle=angle, skip_num_points=0, verbose=False)
+    assert_array_almost_equal(gradex, gradbh, decimal=5)
```
Wouldn't be possible to write a similar test that uses the public API (the TSNE class) to serve as a non-regression test?

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -262,3 +422,126 @@ def test_chebyshev_metric():
+    assert_array_almost_equal(gradex, gradbh, decimal=5)
+
+
+
```
PEP8: 2 blank lines for top level blocks.

amueller commented:
I think at least one person should review the math before merge ;)
Not sure I'll get to it.

On 01/13/2015 10:32 AM, Olivier Grisel wrote:
>
>     @ogrisel <https://github.com/ogrisel> did you review the math?
>
> Not yet.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/scikit-learn/scikit-learn/pull/4025#issuecomment-69762120>.
>

cemoody commented:
> I think at least one person should review the math before merge ;)
> Not sure I'll get to it.

To whoever reviews the math: if you believe the tree was constructed consistently (and there are decent unit tests asserting this) then you have two main 'math' sections to review: the positive and negative gradients. The calculation of the input probability `P(i | j)` is nearly the same as before, except conducted over nearest neighbors instead of all points.

The exact gradient is: ![exact gradient](http://i.imgur.com/Un7craK.png)

And the Barnes-Hut gradient is: ![bh gradient](http://i.imgur.com/7UZwHLv.png)

The first term is `O(uN)` and is computed in `compute_gradient_positive` using `u` nearest neighbors and the second term is `O(N log N)` and is computed in `compute_gradient_negative` using the prebuilt tree.

Carefully reading pages 3 and 4 in the [BH t-SNE paper (pdf)](http://arxiv.org/pdf/1301.3342v2.pdf) should help give context for how these approximations are made. I think by comparing those pages and the code at hand you can review that the math is sane.

We're still slowly working through all of the rest of the comments. Thanks for everyone's diligence!

ogrisel commented:
@cemoody thanks. I will try to find the time to review the math. I am re-reading the paper at the moment.

In the mean time could you please squash the commits of this branch to get a cleaner / simpler history? Then please rebase this branch on the current master and make sure to call `check_fitted(self, 'embedding_')` in the transform method instead of manually raising a value error.

Rebasing without squashing first will be complicated because of intermediary commits that cause conflicts in the past.

arjoly committed:
DOC add performance for all available classifiers

ogrisel commented on doc/modules/manifold.rst:
```diff
@@ -513,12 +515,33 @@ divergence could increase during this phase. Usually it does not have to be
+* The Barnes-Hut implementation only works when the target dimensionality is 3
+  or less. The 2D case is typical when building visualizations.
+* Barnes-Hut only works with dense input data. Sparse data matrices can only be
+  embedded with the exact method.
```
"... or can be approximated by a dense low rank projection for instance using `sklearn.decomposition.TruncatedSVD`."

Please use the `:class:` sphinx reference  format to make have a link to the `TruncatedSVD` API.

cemoody commented:
> Could you add the performance comparison to examples/manifold and make sure that it does not take > much longer than 10^1 seconds (n_samples=1000-2000)?

Hey @AlexanderFabisch, there's already a [t-SNE example for the digits data](http://scikit-learn.org/stable/auto_examples/manifold/plot_lle_digits.html) that takes about ~20sec. This PR makes Barnes-Hut the default, so that time should go down a bit when the docs are regenerated. Is that OK? Could we improve that example or make another example out of a separate dataset?

amueller commented:
I think @AlexanderFabisch's idea was to have the plot that you show at the top in the docs. I'm not really sure that it is necessary, as the main messages is "this is faster". As the bh becomes the default, that doesn't really need much explanation imho.

AlexanderFabisch commented:
> Could you add the performance comparison to examples/manifold and make sure that it does not take > much longer than 10^1 seconds (n_samples=1000-2000)?
> 
> Hey @AlexanderFabisch, there's already a t-SNE example for the digits data that takes about ~20sec. This PR makes Barnes-Hut the default, so that time should go down a bit when the docs are regenerated. Is that OK? Could we improve that example or make another example out of a separate dataset?

That's OK.

cemoody commented:
> That's OK. We don't really need a performance comparison.

I ended up making an example for Olivetti faces using t-SNE and variations on Barnes-Hut / exact, PCA as preprocessing, and PCA as initialization. You can see it in `examples/manifold/plot_tsne.py`. They all have timings displayed, so that should help understanding performance, and should help viewers discover preprocessing with PCA / SVD as a viable path, and discover the other t-SNE options.

amueller commented:
sweet :)

AlexanderFabisch commented on examples/manifold/plot_tsne.py:
```diff
@@ -0,0 +1,122 @@
+dt = t1 - t0
+
+
+t = ("Barnes-Hut t-SNE Visualization of MNIST digits\n" +
```
MNIST digits? :)

cemoody replied:
I must be getting sleepy :)

AlexanderFabisch commented:
> I ended up making an example for Olivetti faces using t-SNE and variations on Barnes-Hut / exact, PCA as preprocessing, and PCA as initialization. You can see it in examples/manifold/plot_tsne.py. They all have timings displayed, so that should help understanding performance, and should help viewers discover preprocessing with PCA / SVD as a viable path, and discover the other t-SNE options.

That's a good idea :)

cemoody committed:
FEAT Barnes-Hut t-SNE

Renamed example file & added comments

FEAT Barnes-Hut t-SNE

Vectorized the math calls -- 3x faster on neg grad

Draft of n-dimensional barnes-hut; failing tests

Tests pass, error being computed in cython

Added error calculation to pos gradient

Changed cases where points are very close to be clearer

Clearer variable names

cemoody commented:
OK, I believe @nicktrav and I have addressed everyone's comments up through this point. While we wait for the math review, let us know what else we can do!

Briefly:
- Added example showing preprocessing with PCA, initialization with PCA, the exact method, and random init, including performance
- Lots of documentation (narrative, examples, and docstrings) added
- Lots of typos and small PEP8 & PEP257 & style fixes
- Tests covering that when angle=0 BH is equal to exact
- In Cython, a small changes to fewer conditional jumps in for loops
- Renaming theta -> angle
- Cleaning up gradient descent arguments
- Fixed tests that were failing 1 in 100 times due to uninitialized arrays
- Fixed failing doctests
- Squashed & cleaned the commit history 

We're also waiting for permission from out comms department to allow a testimonials PR... so that's forthcoming & separate.

GaelVaroquaux commented on examples/manifold/plot_tsne.py:
```diff
@@ -0,0 +1,122 @@
+# Authors: Christopher Erick Moody <chrisemoody@gmail.com>
```
The example is lacking a docstring.

It should be a somewhat complete chunk of text, with a title (the title is used in the gallery) that describes the purpose of the example, and what we can learn from it. And it should come first, before the author line.

GaelVaroquaux commented on examples/manifold/plot_tsne.py:
```diff
@@ -0,0 +1,122 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from matplotlib import offsetbox
+from sklearn.decomposition import RandomizedPCA
```
All sklearn imports should be group in successive lines.

GaelVaroquaux commented on examples/manifold/plot_tsne.py:
```diff
@@ -0,0 +1,122 @@
+dt = t1 - t0
+
+
+t = ("Barnes-Hut t-SNE Visualization of Olivetti Faces digits\n" +
```
I think that "t" is a somewhat short and misleading name for "title". Especially in a context where "t0" and "t1" mean times. :)

GaelVaroquaux commented on examples/manifold/plot_tsne.py:
```diff
@@ -0,0 +1,122 @@
+
+
+t = ("Barnes-Hut t-SNE Visualization of Olivetti Faces digits\n" +
+     "in %1.1f seconds with random initialization and exact method")
```
Am I right in thinking that this title is misleading: for me the Barnes-Hut method what _not_ the exact one.

ogrisel replied:
This should be "Exact t-SNE visualization..." (and just drop the "and exact method" at the end).

ogrisel replied:
Also the digits in "Olivetti Faces digits" seems wrong. I would just use "Olivetti faces". This problem is repeated several times in this file.

cemoody commented:
Thanks @ogrisel + @GaelVaroquaux for the fixes on `plot_tsne.py`. They should be incorporated now. Sorry that so many were typos :)

This PR should have addressed everything up to this point now. I think only the math review is left?

AlexanderFabisch commented on doc/modules/manifold.rst:
```diff
@@ -492,17 +517,28 @@ and select the embedding with the lowest KL divergence.
 
 Optimizing the KL divergence can be a little bit tricky sometimes. There are
-three parameters that control the optimization of t-SNE:
+three parameters that control the optimization of t-SNE and therefore possibly
```
three -> five

ogrisel commented:
The example `plot_tsne.py` could be renamed `plot_tsne_faces.py` to be more explicit. I think it would also be interesting to take one of the tsne on faces plot for inclusion as figure in the narrative documentation in addition to the digits plot.

ogrisel commented on examples/manifold/plot_tsne.py:
```diff
@@ -0,0 +1,137 @@
+time_end = time()
+dt = time_end - time_start
+
+
```
You should create one figure for each plot as done in the `plot_lle_digits.py` example script. Otherwise the aggregated plot is too big to be included in the website.

ogrisel commented on examples/manifold/plot_tsne.py:
```diff
@@ -0,0 +1,137 @@
+f, axs = plt.subplots(2, 2, figsize=(30, 30))
+
+#----------------------------------------------------------------------
+# Fit t-SNE on all 4096 dimensions with random initialization
```
This is not correct: the data is reduced to 50 with `RandomizedPCA` here.

ogrisel commented on examples/manifold/plot_tsne.py:
```diff
@@ -0,0 +1,137 @@
+
+
+title = ("Barnes-Hut t-SNE Visualization of Olivetti Faces\n" +
+         "in %1.1f seconds & initialized embedding randomly")
```
The title should make it explicit that this is on 50 dim reduced data.

ogrisel commented:
I have to run now, but I still want to review the actual implementation. I plan to do that on Friday.

cemoody commented:
Thanks @ogrisel and @AlexanderFabisch. Duly updated the PR to make the latest round of suggested fixes.

cemoody commented:
Hi folks, it's been about a week and a half since the last comment. Let me know if there's anything I can do to help!

amueller commented:
Sorry about the delay. I've been trying to get some bug fixes in before the release. I hope I have time to have a look next week.

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -262,3 +422,128 @@ def test_chebyshev_metric():
+    # check the case where points are actually different
+    Xs.append(np.array([[1, 2], [3, 4]]))
+    # check the case where points are the same on X axis
+    Xs.append(np.array([[-9.368728, 10.264389], [-9.368728, 11.264389]]))
```
Why some many digits? Wouldn't the test be as complete with just `[-1., 2.], [-1., 3.]`, for instance?

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -262,3 +422,128 @@ def test_chebyshev_metric():
+    # check the case where points are the same on X axis
+    Xs.append(np.array([[-9.368728, 10.264389], [-9.368728, 11.264389]]))
+    # check the case where points are arbitraryily close on X axis
+    Xs.append(np.array([[-9.368728, 10.264389], [-9.368761, 11.264389]]))
```
Similar comment, here, it would be easier to follow with points such as:

`[-1.00001, 2.], [-1.00002, 3.]`

ogrisel commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -262,3 +422,128 @@ def test_chebyshev_metric():
+
+    # check the case where points are arbitraryily close on both axes
+    # close to machine epsilon - x axis
+    Xs.append(np.array([[1, 0.0003817754041], [2, 0.0003817753750]]))
```
I think the `dtype` of this arrays should be fixed to `dtype=np.float64` so that the test can be expected to behave consistently across platforms (operating systems and Python architectures).

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # Number of dimensions in the ouput
+    int dimension
+    # Total number of cells
+    long num_cells
```
scikit-learn naming conventions would dictate `n_cells` for this variable.

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # Total number of cells
+    long num_cells
+    # Total number of particles
+    long num_part
```
`n_particules` would be more explicit.

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+
+cdef struct Node:
+    # Keep track of the center of mass
+    float[3] cum_com
```
what does `cum` stand for? "cumulative"? I think this field could be renamed `center_of_mass` or `cumulative_center_of_mass` if there is an ambiguity. Or just `center` if space is an issue (but it does not seem to be the case.

ogrisel replied:
Maybe `barycenter` is better.

cemoody replied:
As an ex-astrophysicist, I approve :)

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    float[3] c
+    # The width of this node -- used to calculate the opening
+    # angle. Equal to width = re - le
+    float[3] w
```
`width`

ogrisel replied:
Actually I have the impression that most of the time `width / 2.0` is actually needed. Wouldn't it make sense to directly store `half_width = width / 2.0` on the node structure?

cemoody replied:
The reason you see `width / 2.0` is because cells split themselves in to two cells (per dimension) recursively. Carrying around a half-width would make that a lot more confusing (has that width been split, or do I need to split again as I subdivide a cell?)

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # Left edge of this node, normalized to [0,1]
+    float[3] le
+    # The center of this node, equal to le + w/2.0
+    float[3] c
```
`center`

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # Holds a pointer to the root node
+    Node* root_node 
+    # Number of dimensions in the ouput
+    int dimension
```
`n_dimensions` to make it explicit that this is a number of dimensions (vs directions for instance).

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    double CLOCKS_PER_SEC
+
+
+cdef struct Node:
```
Maybe this should be renamed `QuadNode` to make it explicit that this is a node in a (point) quadtree.

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # Spit out diagnostic information?
+    int verbose
+
+cdef Tree* init_tree(float[:] left_edge, float[:] width, int dimension, 
```
I would name this class `QuadTree` to be more explicit.

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # Pointer to the tree this node belongs too
+    Tree* tree
+
+cdef struct Tree:
```
I would name this structure `QuadTree` to be more explicit.

ogrisel replied:
Actually if I understand correctly, the code works for `dimension <= 3` and if `dimension == 3` this an octree. Maybe we should rename the structure `QuadOctree` and `QuadOctNode` and make it explicit in a comment that this implements a quadtree if `dimension == 2` and an octree if `dimension == 3`.

cemoody replied:
I've now removed the dimension restriction (though I need to update the documentation)... so it can be `dimension == 2` or `dimension == 6` in principle. Should we stick to Node and Tree?

amueller replied:
I would rather not have just "Tree". There are already to many classes in scikit-learn called "tree" (in the tree module and the neighbors module)

ogrisel commented on sklearn/manifold/t_sne.py:
```diff
@@ -102,26 +156,159 @@ def _kl_divergence(params, P, alpha, n_samples, n_components):
+    else:
+        sP = P.astype(np.float32)
+
+    dimension = X_embedded.shape[1]
```
This is always `n_components` right? Let's reuse that variable instead.

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # Default to leaf until we add particles
+    int is_leaf
+    # Keep pointers to the child nodes
+    Node *children[2][2][2]
```
I would add a comment to explain that for a quad tree, only the first element of the last dimension is used, while for octree both elements of the 3 dimensions are used.

ogrisel commented:
Is the quad tree implementation supposed to work for `n_components==1`? If not I think we should raise an explicit error message when `method='barnes_hut'` (as done for `n_components > 3`).

Apparently there is also an existing bug for that case with the exact method:

```
>>> X = np.random.rand(100, 1000)
>>> TSNE(n_components=1, method='exact').fit(X)
/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py:146: RuntimeWarning: divide by zero encountered in true_divide
  n /= degrees_of_freedom
/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py:148: RuntimeWarning: invalid value encountered in true_divide
  Q = np.maximum(n / (2.0 * np.sum(n)), MACHINE_EPSILON)
Traceback (most recent call last):
  File "<ipython-input-12-ed8483c17c81>", line 1, in <module>
    TSNE(n_components=1, method='exact').fit(X)
  File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 864, in fit
    self.fit_transform(X)
  File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 849, in fit_transform
    embedding = self._fit(X)
  File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 760, in _fit
    skip_num_points=skip_num_points)
  File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 811, in _tsne
    params, error, it = _gradient_descent(obj_func, params, **opt_args)
  File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 382, in _gradient_descent
    new_error, grad = objective(p, *args, **kwargs)
  File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 162, in _kl_divergence
    c = 2.0 * (degrees_of_freedom + 1.0) / degrees_of_freedom
ZeroDivisionError: float division by zero
```

I don't see why `n_components=1` would be forbidden for the exact method so I think this is a bug.

AlexanderFabisch commented:
This bug has been fixed in master
Am 30.01.2015 16:29 schrieb "Olivier Grisel" <notifications@github.com>:

> Is the quad tree implementation supposed to work for n_components==1? If
> not I think we should raise an explicit error message when
> method='barnes_hut' (as done for n_components > 3).
>
> Apparently there is also an existing bug for that case with the exact
> method:
>
> >>> X = np.random.rand(100, 1000)
> >>> TSNE(n_components=1, method='exact').fit(X)
> /Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py:146: RuntimeWarning: divide by zero encountered in true_divide
>   n /= degrees_of_freedom
> /Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py:148: RuntimeWarning: invalid value encountered in true_divide
>   Q = np.maximum(n / (2.0 * np.sum(n)), MACHINE_EPSILON)
> Traceback (most recent call last):
>   File "<ipython-input-12-ed8483c17c81>", line 1, in <module>
>     TSNE(n_components=1, method='exact').fit(X)
>   File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 864, in fit
>     self.fit_transform(X)
>   File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 849, in fit_transform
>     embedding = self._fit(X)
>   File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 760, in _fit
>     skip_num_points=skip_num_points)
>   File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 811, in _tsne
>     params, error, it = _gradient_descent(obj_func, params, **opt_args)
>   File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 382, in _gradient_descent
>     new_error, grad = objective(p, *args, **kwargs)
>   File "/Users/ogrisel/code/scikit-learn/sklearn/manifold/t_sne.py", line 162, in _kl_divergence
>     c = 2.0 * (degrees_of_freedom + 1.0) / degrees_of_freedom
> ZeroDivisionError: float division by zero
>
> I don't see why n_components=1 would be forbidden for the exact method so
> I think this is a bug.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4025#issuecomment-72218095>
> .
>

ogrisel commented:
I think the current test coverage of the quad tree implementation is too weak. I agree it's painful to write tests for Cython datastructures in Cython itself. I think I would rather like to expose a subset of the internal quadtree API just for testing purpose. For instance to build a tree from a set of 2D or 3D points passed as 2D numpy array with shape `(n_points, n_dimensions)` and then introspecting the properties of the leaf cells.

For instance a test set could be in 2D:

```python
X = np.array([
    [0, 0],
    [0.25, 0.75],
    [0.75, 0.25],
    [1, 0],
    [0, 1],
    [-1, -1],
])
```

Then inspecting that we have 3 large leaf cells with `width==1.0` for the negative quadrants and smaller cells for the positive quadrant.

ogrisel commented:
> This bug has been fixed in master

Great then we need to rebase this PR and add a good error message for the case `n_components=1, method='barnes_hut'`.

ogrisel commented on examples/manifold/plot_tsne_faces.py:
```diff
@@ -0,0 +1,148 @@
+In addition to varying method & init, the script also compares preprocessing
+the input data by reducing the dimensionality via PCA and also varying the
+perplexity.
+
```
I think this example is too crowded / long and we don't get the intended message of the example. I would rather simplify it by just demonstrating the cases with the Barnes Hut method and the 3 following configurations:

- random init on raw features
- PCA init on raw features
- PCA init on PCA reduced features (via a pipeline).

The in the text description of the example I would tell that:

- all the three configurations yield similar arrangements of the data,
- PCA-reduced features is interesting to speed up the analysis while preserving most of the pairwise distances of the original feature space. In particular the BallTree data structure used to compute the neigborhoods in the original feature space is known to work faster than brute force only for lower dimensional space (e.g. less than a couple hundreds of dimensions).
- PCA init is interesting to get global arrangement of the groups that are slightly more consistent with the long range dependencies between groups (although I am not 100% sure about this so maybe we should not mention that all).

cemoody replied:
Great! Will do.

I think it's fine to say the initializing with PCA preserved more of the global structure -- LvdM mentions this in his papers.

amueller replied:
I would also put all plots in a single figure so they are side by side

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # Keep track of the center of mass
+    float[3] cum_com
+    # If this is a leaf, the position of the particle within this leaf 
+    float[3] cur_pos
```
`leaf_particule_position`?

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    float[3] cur_pos
+    # The number of particles including all 
+    # nodes below this one
+    long cum_size
```
`n_particules`

ogrisel replied:
Or `n_points`: we need to decide between "particule" and "point" and be consistent in the name throughout the quad tree code.

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # Number of particles at this node
+    long size
+    # Index of the particle at this node
+    long point_index
```
This is defined only for non-empty leaf nodes, right? The comment could make that explicit.

ogrisel replied:
Or  `particule_index`: we need to decide between "particule" and "point" and be consistent in the name throughout the quad tree code.

cemoody replied:
I'll pick 'point' (as in data point) and try to stick to it.

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # And each subdivision adds 1 to the level
+    long level
+    # Left edge of this node, normalized to [0,1]
+    float[3] le
```
`left_edge` is a better name.

ogrisel commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,682 @@
+    # level = 0 is the root node
+    # And each subdivision adds 1 to the level
+    long level
+    # Left edge of this node, normalized to [0,1]
```
By reading the code I think that this is not normalized to the range [0,  1]. Is this comment referring to and older version of the code?

ogrisel commented:
I am still not done reading the Cython source code. I hope I will find some time next week (or maybe during a boring talk this WE at FOSDEM 2015).

mjbommar committed:
Adding fix for issue #4297, isotonic infinite loop

gcr commented:
Thanks for making this fork! This will benefit me *immensely*.

I have a few questions though, mostly coming from a **memory use** standpoint. Currently, the `_joint_probabilities_nn` function calls `_utils._binary_search_perplexity`, which allocates an **NxN distance matrix.** The `_joint_probabilities_nn` function then does more numpy-like calculations which allocates several more copies of this distance matrix in memory!

- Doesn't this imply the worst-case time complexity *must* be at *least* O(n^2), since each distance matrix element must be filled?

- What if I want to embed millions of points? To embed 1e6 points, storing the distance matrix would require terabytes of memory. Has anyone tested this fork on millions of points? In my experience, embedding 10,000 points even with this fork requires >4GB of memory for the intermediate computation.

In the [Barnes-Hut paper](http://arxiv.org/pdf/1301.3342v2.pdf), Laurens embeds all of TIMIT which contains >1,000,000 points. How is he storing that `P` distance matrix? Is he just `mmap(3)`ping a terabyte of data onto disk somehow?

ogrisel commented:
This is a very good remark. We should definitely avoid computing the full `n_samples x n_samples` `distances` matrix for the Barnes-Hut t-SNE method but instead chunk X into minibatches of `batch_size=1000` samples at a time that are fed to the ball-tree to query one batch at time and only incrementally store the distances to the neighbors in a `distances_to_neighbors` sparse matrix (typically a COO or a CSR).

Both `_utils._binary_search_perplexity` and `_joint_probabilities_nn` should be updated to deal with that COO or CSR matrix instead of the current `(distances, neighbors)` pair of arrays.

If you don't want to mess with scipy sparse COO or CSR datastructure it might be even easier to make `distances` a float array with shape `(n_samples, n_neighbors)` and make `neighbors` an integer array with shape `(n_samples, n_neighbors)` with the indices of the n_neighbors nearest neighbors for each sample in the dataset.

In both cases it should remove the O(N^2) space complexity.

WDYT @cemoody ? If the above outline is no clear enough let me know. I have not found the time to do the full review as originally intended but plan to do so after the 0.16 beta is out. If you no longer have time to work on that PR anymore let us know, someone else can take over from here.

cemoody commented:
Thanks @gcr, and @ogrisel! The memory complexity for the NxN step is indeed pretty bad. That distance calculation is used to compute the perplexity (~number of neighbors) for every data point. To do that you don't really need to compute all NxN distances, just some small fraction of the nearest. LvdM actually only keeps around u * N neighbors for every point. 

I have a branch (https://github.com/cemoody/scikit-learn/blob/cemoody/bhtsne02) that systematically addresses all of these issues (except for using sparse matrices), reduces memory in computing the cost function too (that's a hot spot too, not just the distance matrix), removes the limit of using 2-3 dimensions for BH t-SNE, and vectorizes a few spots to get a 3-4x performance gain. 

I think of this PR as adding a bunch of features where the math, theory and approximations should be carefully vetted (thank you @ogrisel for reviewing when you can!). After this gets approved, I can submit a second PR that adds a ton of performance (both speed and memory) enhancements. I didn't want to prematurely optimize a bunch of stuff and throw a wrench into things while we're still reviewing this one. I think that makes it easier, but if you guys think it's wiser to merge that into one PR, I'm fine with that too. One PR or two?

Also, apologies that I haven't responded to the last round -- I started at Stitch Fix a few weeks ago and should have the time to address this soon :)

ogrisel commented:
Glad to hear that @cemoody, I think we should rather review directly the optimized version. Or at least a version that includes the fixes to get the `u * N` memory usage scalability. And congrats on your new job :)

amueller committed:
Fix rebase conflict

amueller commented:
+1 on doing it in one PR. This is on my todo for next week (for whatever it's worth).

amueller committed:
Move dev branch to 0.17

cemoody commented:
Great! I'll merge the two PRs and try to respond to all of the suggestions above.

cemoody committed:
Merge branch 'master' into cemoody/bhtsne

Conflicts:
	sklearn/manifold/t_sne.py
	sklearn/manifold/tests/test_t_sne.py

cemoody commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,829 @@
+    l = <long*> malloc(sizeof(long))
+    neg_force= <float*> malloc(sizeof(float) * n_dimensions)
+
+    for i in range(start, stop):
```
Incidentally, this `for` loop is ripe for parallelization. It is the most expensive part of the gradient calculation. The memory access is opportune as well: the tree is only read from, no writing. And there's only a few small temporary variables in the body of the `for` loop, followed by never-overlapping writes to a single row of a force matrix that can be done in any order. I'm not an expert at multiprocessing in Cython, and my initial attempt with `prange` didn't yield any improvements. If there's an easy way to parallelize this for loop, could be a quick win and an easy way to be faster than the reference implementation :)

amueller commented:
If possible it would be good if you could avoid merging master into your branch. Instead you should ``git rebase master``. In case that gets messy, you can squash your commits before with ``git rebase -i``.

amueller commented on sklearn/manifold/setup.py:
```diff
@@ -14,6 +16,16 @@ def configuration(parent_package="", top_path=None):
+    eca.append("-O4")
+    config.add_extension("_barnes_hut_tsne",
+                         libraries=cblas_libs,
+                         sources=["_barnes_hut_tsne.c"],
```
I think you need to add utils here. What was the motivation of putting one function in a different file? Do you think it will be more widely useful?

amueller replied:
Never mind, it looks like that was there before and it gets correctly built.

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -433,34 +690,73 @@ def fit(self, X, y=None):
+                neighbors_nn = np.argsort(distances, axis=1)[:, :k]
+            else:
+                # Find the nearest neighbors for every point
+                bt = BallTree(X)
```
Although a little bit of validation overhead is added, I would rather see the public `NearestNeighbors().kneighbors` abstraction used here. This may, for example, allow us to support approximate nearest neighbors methods in the future (and could perhaps support sparse input now).

ogrisel replied:
+1

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -433,34 +690,73 @@ def fit(self, X, y=None):
+                # In the event that we have very small # of points
+                # set the neighbors to n - 1
+                distances_nn, neighbors_nn = bt.query(X, k=k+1)
+                neighbors_nn = neighbors_nn[:, 1:]
```
This won't be correct if a data point is duplicated. A correct version of this functionality is now provided by `NearestNeighbors(...).fit(X).kneighbors()` (don't pass `X` to `kneighbors` and it will query with the indexed data, excluding the query point from the results).

ogrisel commented:
I seems that something wrong happened during the rebase. Could you please get rid of the spurious commits that ended up in this branch?

amueller commented:
There is a linking issue with blas:
https://travis-ci.org/scikit-learn/scikit-learn/jobs/53473622#L1126

amueller commented:
I thought the blas linking was caused by not specifying the library path (other setup.py files use ``**blas_info`` that contains ``library_dirs``) but that doesn't seem to be the issue.
Any ideas anyone?

amueller commented on examples/manifold/plot_tsne_faces.py:
```diff
@@ -0,0 +1,148 @@
+"""
+===================================
```
fix underline / overline length

amueller committed:
add necessary blas files.

amueller commented:
Ah, the blas files were missing. PR here: https://github.com/cemoody/scikit-learn/pull/9

amueller commented on sklearn/manifold/t_sne.py:
```diff
@@ -521,5 +842,57 @@ def fit_transform(self, X, y=None):
+        self.embedding_ = embedding
         return self.embedding_
+
+    def fit(self, X):
```
Can you maybe remove this method as was done in master? I find it confusing and it doesn't add anything.
Also, ``fit`` needs to return ``self``

amueller replied:
``fit`` also needs to take an optional ``y=None`` even if it is ignored.

amueller commented on sklearn/manifold/t_sne.py:
```diff
@@ -507,8 +827,9 @@ def _tsne(self, P, alpha, n_samples, random_state, X_embedded=None):
 
-    def fit_transform(self, X, y=None):
-        """Transform X to the embedded space.
+    def fit_transform(self, X):
```
``fit_transform`` needs to accept an optional ``y=None`` even if it is ignored.

cemoody committed:
Merge pull request #9 from amueller/tsne_review

add necessary blas files.

amueller commented on sklearn/manifold/t_sne.py:
```diff
@@ -477,29 +773,53 @@ def _tsne(self, P, alpha, n_samples, random_state, X_embedded=None):
             print("[t-SNE] Error after %d iterations with early "
                   "exaggeration: %f" % (it + 1, error))
+        # Save the final number of iterations
+        self.n_iter_final = it
```
This is number of iterations actually performed, right? Then it should have an underscore at the end.

ogrisel replied:
I actually it should be named `n_iter_` to be consistent with the conventions of the project.

amueller commented:
I think there should be a test comparing the exact and approximate method at least on some very small dataset (so that it runs in ~1s if possible). Currently, I am not sure if the exact method is tested well any more.
It might make sense to refactor the tests to run them with both algorithms.

amueller commented on sklearn/manifold/_utils.pyx:
```diff
@@ -2,33 +2,36 @@ from libc cimport math
+    """Binary search for sigmas of conditional Gaussians. 
+    
+    This approximation reduces the computational complexity from O(N^2) to
+    O(uN). See the exact method '_binary_search_perplexity' for more details.
```
where is that?

amueller replied:
Does ``neighbors = None`` correspond to the eact method?

amueller commented on doc/modules/manifold.rst:
```diff
@@ -479,12 +479,37 @@ t-distributed Stochastic Neighbor Embedding (t-SNE)
+* Revealing data that lie in multiple, different, manifolds or clusters
+* Reducing the tendency to crowd points together at the center
+
+While Isomap, LLE and variants are best suited to unfold a single continuous
```
I didn't check but that sounds quite verbatim from the paper. If it is, we should probably make that explicit.

amueller commented:
Where does the name ``angle`` come from? It is not in the paper, is it? Maybe ``approximation`` or ``approximation_quality`` or something would be better?

amueller commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,829 @@
+    long level
+    # Left edge of this node
+    float* left_edge
+    # The center of this node, equal to le + w/2.0
```
le == left_edge, I guess?

amueller commented on sklearn/manifold/_barnes_hut_tsne.pyx:
```diff
@@ -0,0 +1,829 @@
+    # The center of this node, equal to le + w/2.0
+    float* center
+    # The width of this node -- used to calculate the opening
+    # angle. Equal to width = re - le
```
re = right_edge, le = left_edge.

amueller commented:
@cemoody did you have time to merge the two PRs? That would be really great!

amueller commented:
It looks to me like https://github.com/cemoody/scikit-learn/tree/cemoody/bhtsne02 still needs n**2 memory as it computes the full distance matrix. Is that not the case?

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -433,34 +690,73 @@ def fit(self, X, y=None):
+                # set the neighbors to n - 1
+                distances_nn, neighbors_nn = bt.query(X, k=k+1)
+                neighbors_nn = neighbors_nn[:, 1:]
+            P = _joint_probabilities_nn(distances, neighbors_nn,
```
Along the lines of @amueller's remark, should `distances` here be `distances_nn`?

cemoody commented:
How folks, I just wanted to update everyone that I'm still working on this, and am especially focused on including CSR sparse arrays for better memory optimization and cleaner code. Sorry for the long bout of silence!

amueller commented:
Thanks for letting us know. It would be great if you could address our comments above. I'm very excited about your contribution!

amueller commented:
Is the CRS matrices for dense input or sparse input? I tried running MNIST on your feature branch but there still seems to be a quadratic memory requirement.

cemoody commented:
@amueller Scipy's CSR matrices are just for computing the (n_samples, n_samples) `distances` matrix in a sparse way. I'm not (yet) trying to support a sparse input dataset for `method='barnes-hut'`, just trying to reduce the memory pressure in intermediate stages of the method. Roughly, a sketch of the basic t-SNE algorithm goes like this:

1. Start with input dataset `(n_samples, n_features)`
2. Build a sample-sample distance matrix. `O(n_sample^2)` memory and `O(n_sample^2)` performance.
3. Build `P(i|j)` matrix where `i` & `j` are the input data. `O(n_sample^2)` memory and `O(n_sample^2)` performance.
4. Do gradient descent such that the embedding minimizes KL divergence with the `P(i|j)` of the input. 

What LvdM implements does this:
1. Start with `(n_samples, n_features)`
2. Build a sample-sample distance matrix. Only do this for nearest neighbors. Use a sparse matrix. Now we pay the one-time upfront cost of building a neighbor tree, but memory performance is then `O(u * n_samples)` since we keep `u` neighbors of every sample.
3. Build `P(i|j)` matrix where `i` & `j` are the input data. Again only define `P(i|j)` sparsely, when `i` & `j` are neighbors, else `P(i|j)` is zero. 
4. Do gradient descent such that the embedding minimizes KL divergence with `P(i|j)` of the input. We get a linear speed boost from a sparse `P(i|j)`, since we don't scan over all elements of `P(i|j)`. This is OK theoretically, what the paper recommends, and what the reference code does.

On theoretical grounds, keeping `P(i|j)` a sparse matrix is quite attractive since the KL divergence is highly asymmetric. This means that high probabilities matter a lot, and low `P(i|j)` are ignorable even when 

In terms of code change, we basically already had this implemented as a dense `P(i|j)` array and accompanying `neighbors` array. So we were taking up a ton of memory, but weren't actually using anything but the neighbors. We just need to switch to a fully sparse `P(i|j)` and then the filled-in indices will be equivalent to the current `neighbors` array. 

Apologies for the wall of text, and for likely over-answering :)

amueller commented:
Hehe, better too much then to little detail. Let us know if you make some progress :)

amueller commented:
I tried to rebase in #4887.
Currently on my box gcc fails:
```
sklearn/manifold/_barnes_hut_tsne.c: In function ‘__pyx_pf_7sklearn_8manifold_16_barnes_hut_tsne_2gradient’:
sklearn/manifold/_barnes_hut_tsne.c:7159:5: error: format not a string literal and no format arguments [-Werror=format-security]
     printf(__pyx_k_t_SNE_Computing_gradient);
```
I have no idea what is happening here.

ogrisel commented:
I don't get the gcc failure on my box. I am running gcc 4.8.2-19ubuntu1. Maybe @larsmans has an idea? It might be a bug in Cython.

Other than that I get the same tests errors as travis is currently reporting in #4887.

amueller commented:
I'm running ``gcc (Ubuntu 4.9.2-10ubuntu13) 4.9.2``. And I get it with this branch and my rebase, and with the checked-in cython 0.21 and my 0.22 :-/ Maybe I'm doing something silly.

amueller commented:
I think the failures in #4887 are all input validation. I think we should merge this even without the memory improvement as it gives a 10x speedup on decent sized datasets.

ogrisel commented:
I have no issue building this branch with clang on OSX. Will try with the latest version of gcc with docker.

ogrisel commented:
I can reproduce your compile error with gcc 4.9.2-10ubuntu13 from ubuntu 15.04. I looks like a Cython bug: when calling `printf("some string\n")` in Cython, it creates a local string variable that is not a `const` and that causes the `-Werror=security-format` to cause a compiler error with gcc 4.9.

When calling `printf("some string with a formatted value: %d\n", i)` then the generated C code has a `const char *` as first argument and everything is fine.

ogrisel commented:
The following hack seems to work:

```
diff --git a/sklearn/manifold/_barnes_hut_tsne.pyx b/sklearn/manifold/_barnes_hut_tsne.pyx
index a5579b5..3978c80 100644
--- a/sklearn/manifold/_barnes_hut_tsne.pyx
+++ b/sklearn/manifold/_barnes_hut_tsne.pyx
@@ -14,6 +14,8 @@ from libc.math cimport sqrt, log
 cimport numpy as np
 import numpy as np

+cdef char* EMPTY_STRING = ""
+
 cdef extern from "math.h":
     float fabsf(float x) nogil

@@ -751,13 +753,19 @@ def gradient(float[:,:] pij_input,
     err = insert_many(qt, pos_output)
     assert err == 0, "[t-SNE] Insertion failed"
     if verbose > 10:
-        printf("[t-SNE] Computing gradient\n")
+        # XXX: format hack to workaround lack of `const char *` type
+        # in the generated C code that triggers error with gcc 4.9
+        # and -Werror=format-security
+        printf("[t-SNE] Computing gradient\n%s", EMPTY_STRING)
     sum_Q = compute_gradient(pij_input, pos_output, neighbors, forces,
                              qt.root_node, theta, dof, skip_num_points, -1)
     C = compute_error(pij_input, pos_output, neighbors, sum_Q, n_dimensions,
                       verbose)
     if verbose > 10:
-        printf("[t-SNE] Checking tree consistency \n")
+        # XXX: format hack to workaround lack of `const char *` type
+        # in the generated C code
+        # and -Werror=format-security
+        printf("[t-SNE] Checking tree consistency\n%s", EMPTY_STRING)
     cdef long count = count_points(qt.root_node, 0)
     m = ("Tree consistency failed: unexpected number of points=%i "
          "at root node=%i" % (count, qt.root_node.cumulative_size))
```

ogrisel commented:
There are other occurrences of the same issue in the file. The above patch does not fix them all.

cemoody commented:
Alas, I was overly optimistic in how much time I'd have to work on this. Can we make a fresh TODO list of what this PR needs?

amueller commented:
@cemoody welcome to the club ;)
Maybe reset it to #4887 and fix the errors.

@ogrisel is the cython bug fixed in newer cython? Should we report?

I think having a sparse matrix variant would be nice but I would be happy to merge without it.

amueller commented:
I added @ogrisel's fix to #4887. However, the ``transform`` method can't work the way it is implemented now. I haven't really looked at it, but it currently assumes that the input dimension is the same as the output dimensions somehow. It concatenates the embedding with the new data. I think it should just concatenate the training data with the new data, right? But then the tests should be tighter.

amueller commented:
Fun api issue: calling ``transform`` will change ``_X_fit``, so calling ``transform`` changes again afterwards will change the results...

amueller commented:
Also, calling transform changes the embedding of the training data. I thought that was supposed to be avoided using ``skip_num_points``.
``transform`` also returns the transform of the training data plus the additional data, which breaks api.
Maybe we should try to merge this without the ``transform``? Or does someone have time to look into it?

zachmayer commented:
Is this project still in progress?  I'd really love to be able to use this new algorithm.

amueller commented:
@zachmayer the algorithm is in master and will be in the imminent release. It uses dense data structures, though, so the memory requirements will be the same as before. There is no-one working on using sparse data structures atm afaik.

zachmayer commented:
I can live with that!  Awesome!

cemoody commented:
Oh, I didn't realize it was in master! That's super awesome & I'm incredibly excited!!! Thanks @amueller for all of your hard work :)

zachmayer commented:
@amueller Can you link to the PR/commit that puts it in master?  Thanks!

cemoody commented:
@zachmayer I think it's at this commit: https://github.com/scikit-learn/scikit-learn/commit/fe09a077da0e6fcf6b92b0281523952d0d4874e0

amueller commented:
This one: https://github.com/scikit-learn/scikit-learn/pull/4887

I left your PR open, as I removed the "transform" method. It changed the training data, which is not really allowed. I think that might not have been what you intended, was it?

jnothman commented:
I assume this should be closed?

jnothman commented:
It can be reopened if that's appropriate.

jamiis commented:
@cemoody @amueller is there still ongoing work to optimize memory consumption occurring in calculating the pairwise distances?

ogrisel commented:
@jamiis AFAIK nobody is working on it at the moment. Do you want to give it a try?

jlorince commented:
In principle, shouldn't there be a way to avoid calculating the full distance matrix? The original author of t-SNE says "Barnes-Hut algorithms are intended to operate on large datasets for which you cannot keep such distance matrices in memory" (see https://github.com/lvdmaaten/bhtsne/issues/7), so I'm wondering if the current approach in the sklearn version (starting by calculating the full distance matrix) is inherently flawed. Though honestly I'm not sure what the correct alternative is....

AlexanderFabisch commented:
I think the best way to implement it is a sparse distance matrix where only the relevant entries are included. However, it is a little bit complicated to handle all special cases correctly, e.g. precomputed distance matrices.

Dringite commented:
Any progress? I'm unable to find an implementation that lets you both work on large datasets and vary the number of iterations.

amueller commented:
Moving the discussion to #7089. PR welcome ;)
